### ISSUE TITLE
General Feedback - Feature Requests, Bugs

### ISSUE THREAD

**Issue by iMrShadow** [Shadowth117/DrSwizzler#1]
Hi! 

I recently discovered this library while looking for unswizlling/swizzling methods for a texture converter (and this is the only one that I know of). So far it has been great using it, I really enjoyed it! I have some requests and I am not sure if I should split into multiple issues or not, please let me know. 

Feature Requests 
- [ ] Wii, WiiU and Xbox One swizzle/deswizzle support 
- [x] Add more [Xbox360 formats](https://github.com/Shadowth117/DrSwizzler/blob/main/Swizzling/Xbox360Deswizzler.cs#L27) (A8, L8 and B8G8R8X8 come to my mind) – (waiting for #3 to be merged)

Bugs 
- [x] Nintendo deswizzling doesn't work, because the [tempBuffer](https://github.com/Shadowth117/DrSwizzler/blob/main/Swizzling/SwitchDeswizzler.cs#L63) is returned rather than the [outBuffer](https://github.com/Shadowth117/DrSwizzler/blob/main/Swizzling/SwitchDeswizzler.cs#L17) (fixed in #2)
- [x] [Small typo](https://github.com/Shadowth117/DrSwizzler/blob/main/Swizzling/Xbox360Deswizzler.cs#L27) - should be Align. (fixed in #2)

I could also pull request if you want.

Thanks!

**Comment by Shadowth117** [comment]
Hey, yeah. Go for it. 
I'm not the best at doing these sorts of algorithms myself, but I wanted to make a lib that's somewhat standardized in case people want to do that. I would love to have support for other platforms, though I'm not sure I'd be the best choice to implement some of those. I am absolutely on board to help manage getting things into this, however. 

I merged what you had in the PR FYI.

**Comment by Shadowth117** [comment]
Something else I'd thought about too was potentially handling layouts for miplevels for a full texturebuffer, though I'm not certain how standardized that is. The origin for the Xbox 360 portion did try to read all mip levels at one point, but I don't think it supported how it worked in the original Dark Souls so MGS V 360 may not have had the smallest mip levels at all (I didn't have the game at the time to test this, myself). If it did read down to 1x1 it would definitely be a different implementation though. 

Basically though, it's just a thought and I don't know enough to be sure if attempting to read the larger layout from texture buffers would be useful and if padding would vary a ton per game.

**Comment by iMrShadow** [comment]
Apologies for not replying, life got into the way.

[ReverseBox](https://github.com/bartlomiejduda/ReverseBox) is a project which has additional algorithms for swizzling and deswizzling, can be useful for adding new platforms or improving current functions.

> Something else I'd thought about too was potentially handling layouts for miplevels for a full texturebuffer, though I'm not certain how standardized that is. The origin for the Xbox 360 portion did try to read all mip levels at one point, but I don't think it supported how it worked in the original Dark Souls so MGS V 360 may not have had the smallest mip levels at all (I didn't have the game at the time to test this, myself). If it did read down to 1x1 it would definitely be a different implementation though.

It could be a nice QoL feature, even though I personally would not use it, and I will explain why.

The way it should work is to split and take all images from a buffer (knowing the width, height, depth, format, mips and layout) and apply swizzling/deswizzling to each of them. The problems come from splitting the texture into multiple images:
1. The texture can have only a couple of mips.
2. It can be a different layout - Arraymap, Volumemap, Cubemap, Array cubemap.
3. Different container formats. One example is KTX, which is a popular texture container. Unlike DDS, it stores its mips from smallest to largest.

As I said - nice QoL feature, but I would personally apply the functions to a specific mip/image, like I have done in my texture tool for better control. 

How do you want to implement it?

**Comment by Shadowth117** [comment]
> Apologies for not replying, life got into the way.
> As I said - nice QoL feature, but I would personally apply the functions to a specific mip/image, like I have done in my texture tool for better control.
> 
> How do you want to implement it?

All good. I should really check notifs here more, myself.

Uh, yeah you bring up a lot of good points there and a lot of that is part of why I didn't do it. I figure if some dev is really insane they could pretty much do what they wanted with miplevels and the like anyways. 

For now though, I have gone and merged that PR you put up the other day. It does look like it handles a lot of this more sensibly. I'm definitely not the biggest expert on all of this, but I've been trying to learn as I go along and hopefully this all will be valuable to others wanting to play about with it all too.

**Comment by CyB1K** [comment]
Heya first of all thanks so much for this, seems to be the only tool available everywhere that can swizzle ps4 textures, i dont know if this is still worked on or not but i have an issue when swizzling ps4 textures, it seems drswizzler adds some extra data on the textures randomly which causes an odd rectangle on deswizzled textures. Also seems to always be on bottom right side of the texture

A few examples:

https://i.imgur.com/hhd8Av7.png
https://i.imgur.com/onK2Ubf.png

https://i.imgur.com/8eGIVtk.png
https://i.imgur.com/VNgctXB.png

Thanks and have a nice day

Nevermind, turns out the issue was not swizzling it but exporting the texture, i was using dds tools to export and it was breaking it (i think mipmaps)

**Comment by iMrShadow** [comment]
This is indeed the only Nuget package available that provides PS4 swizzling, but there are many other tools that can deswizzle, like RawTex (Raw Texture Cooker) or [ImageHeat/Reversebox](https://github.com/bartlomiejduda/ImageHeat), which you can check out.

The implementation of the PS4 swizzling algorithm is based on RawTex, which in theory always works on non-even textures (speaking from experience). It does not add any extra data

It's kinda odd that the textures in the screenshots have a single corrupted area. Have you found a solution for them?

**Comment by CyB1K** [comment]
Nope i could not find a way to fix it, this seems related to BC7. I managed to sorta bypass it by forcing uncompressed B8G8R8A8 instead

**Comment by Shadowth117** [comment]
> Nevermind, turns out the issue was not swizzling it but exporting the texture, i was using dds tools to export and it was breaking it (i think mipmaps)

So to be clear on this then, you managed to work it out? 
I will say sometimes gets weird at the mip level so it's important you get pretty exact buffer sizes. You will have cases where you need a larger buffer size than you might expect due to how the algorithm ends up working.
